### PR TITLE
Migrate all Peer communication to common Notification subsystem

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Migrate all Peer communication to common Notification subsystem

## Motivation and Context
Deprecate the use of Admin Peers concept and migrate all peer
communication to Notification subsystem. This finally allows
for a common subsystem for all peer notification in case of
distributed server deployments.

## Regression
No

## How Has This Been Tested?
Generally testing all existing functionality
```
~ mc admin profile
```
```
~ mc admin service status myminio/
```
```
~ mc admin info myminio/
```

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by balamurugana — DISMISSED** [review]

**Comment by harshavardhana** [comment]
Rebased to master please review @balamurugana @vadmeste

**Review by balamurugana — DISMISSED** [review]

**Comment by harshavardhana** [comment]
PR rebased to master please review @balamurugana @vadmeste

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#e7128ed2-3fe5-4999-b4d0-ac059d8a7ea1) |
<a name="e7128ed2-3fe5-4999-b4d0-ac059d8a7ea1"></a>
### 7031-928ee4d/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30006
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 775a5f1dcf0e:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 2 minutes and 53 seconds
{
  "name": "awscli",
  "duration": 70990,
  "function": "delete_bucket\n",
  "status": "FAIL",
  "error": "remove_bucket failed: s3://awscli-mint-test-bucket-27161 An error occurred (BucketNotEmpty) when calling the DeleteBucket operation: The bucket you tried to delete is not empty"
}

Executed 0 out of 13 tests successfully.
```

**Review by balamurugana — DISMISSED** [review]

**Comment by harshavardhana** [comment]
ping @vadmeste

File: cmd/peer-rpc-server.go [hunk 1]
```diff
@@ -249,6 +252,126 @@ func (receiver *peerRPCReceiver) DrivePerfInfo(args *AuthArgs, reply *ServerDriv
...
+		if info.Error != "" {
+			continue
+		}
+		times = append(times, info.Data.Properties.Uptime)
+	}
+
+	// Sort uptimes in chronological order.
+	sort.Sort(uptimes(times))
+
+	// Return the latest time as the uptime.
```
Comment (vadmeste) [inline cmd/peer-rpc-server.go:311]: You mean "return the uptime of the youngest node", right ?

File: cmd/notification.go [hunk 2]
```diff
@@ -171,6 +173,174 @@ func (sys *NotificationSys) LoadCredentials() map[xnet.Host]error {
...
 
+// StartProfiling - start profiling on remote peers, by initiating a remote RPC.
+func (sys *NotificationSys) StartProfiling(profiler string) map[xnet.Host]error {
+	errors := make(map[xnet.Host]error)
+	var wg sync.WaitGroup
+	for addr, client := range sys.peerRPCClientMap {
+		wg.Add(1)
+		go func(addr xnet.Host, client *PeerRPCClient) {
+			defer wg.Done()
+			// Try to load credentials in three attempts.
```
Comment (vadmeste) [inline cmd/notification.go]: I think this comment needs to be fixed, also I am not sure about the necessity of doing three retries

File: cmd/utils.go [hunk 3]
```diff
@@ -208,6 +208,24 @@ func (p profilerWrapper) Path() string {
 	return p.pathFn()
 }
 
+func downloadProfileData() ([]byte, error) {
```
Comment (vadmeste) [inline cmd/utils.go]: 'download' is little misleading, I think stopNGetProfilingData() would be a better name.

File: cmd/admin-handlers.go [hunk 4]
```diff
@@ -359,31 +366,25 @@ func (a adminAPIHandlers) StartProfilingHandler(w http.ResponseWriter, r *http.R
...
-			result.Success = true
-			startProfilingResult[idx] = result
-		}(i, peer)
+	var startProfilingResult []StartProfilingResult
+
+	for host, err := range globalNotificationSys.StartProfiling(profiler) {
+		result := StartProfilingResult{NodeName: host.String()}
+		if err != nil {
+			result.Error = err.Error()
+			continue
```
Comment (vadmeste) [inline cmd/admin-handlers.go]: This is a bug from the old code, we should add result to startProfilingResult list before continuing

File: cmd/admin-handlers.go [hunk 5]
```diff
@@ -242,36 +250,35 @@ func (a adminAPIHandlers) ServerInfoHandler(w http.ResponseWriter, r *http.Reque
...
-				logger.LogIf(ctx, err)
-				reply[idx].Error = err.Error()
-				return
-			}
+	storage := objectAPI.StorageInfo(ctx)
 
-			reply[idx].Data = &serverInfoData
-		}(i, p)
+	serverInfo := globalNotificationSys.ServerInfo(ctx)
+	serverInfo[*thisAddr] = ServerInfo{
```
Comment (vadmeste) [inline cmd/admin-handlers.go]: I am not sure why we are overriding serverInfo contents after calling globalNotificationSys.ServerInfo(ctx), can you add a comment here ?

**Review by vadmeste — COMMENTED** [review]
Some comments

File: cmd/peer-rpc-server.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/peer-rpc-server.go:311]: Yeah latest is youngest..

File: cmd/notification.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/notification.go]: Retries is just to ensure the operation is complete. It is simply a safety mechanism for network errors. We don't have retries in rpc so we have it here

File: cmd/admin-handlers.go (hunk 4 above)
Comment (harshavardhana) [reply cmd/admin-handlers.go]: Alright

File: cmd/admin-handlers.go (hunk 5 above)
Comment (harshavardhana) [reply cmd/admin-handlers.go]: ServerInfo is a map this is just adding local node.

File: cmd/utils.go (hunk 3 above)
Comment (harshavardhana) [reply cmd/utils.go]: Good idea..

**Comment by minio-ops** [comment]
```
Error running mint automation
+ mkdir -p 7031-4ca377a/gopath/src/github.com/minio
+ git clone --quiet https://github.com/minio/minio.git 7031-4ca377a/gopath/src/github.com/minio/minio
++ cd 7031-4ca377a/gopath/src/github.com/minio/minio
++ git remote add minio https://github.com/minio/minio.git
++ cd 7031-4ca377a/gopath/src/github.com/minio/minio
++ git fetch --quiet minio pull/7031/head:pr7031
++ cd 7031-4ca377a/gopath/src/github.com/minio/minio
++ git checkout --quiet pr7031
++ minikube ip
There is a newer version of minikube available (v0.32.0).  Download it here:
https://github.com/kubernetes/minikube/releases/tag/v0.32.0

To disable this notification, run the following:
minikube config set WantUpdateNotification false
+ minikube_ip=192.168.39.88
++ mkdir -p certs
++ cd certs
++ ./cert -host 192.168.39.88
2019/01/06 00:56:12 wrote public.crt
2019/01/06 00:56:12 wrote private.key
+ GOPATH=/mnt/nvme0n1p1/mint-auto/mint-auto/7031-4ca377a/gopath
+ make -C 7031-4ca377a/gopath/src/github.com/minio/minio --quiet
# github.com/minio/minio/cmd
cmd/peer-rpc-server.go:316:15: undefined: downloadProfileData
make: *** [build] Error 2

```

**Comment by harshavardhana** [comment]
PR updated after fixing some review comments - PTAL @balamurugana @vadmeste

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by balamurugana — DISMISSED** [review]

File: cmd/notification.go [hunk 6]
```diff
@@ -171,6 +173,174 @@ func (sys *NotificationSys) LoadCredentials() map[xnet.Host]error {
...
+				}
+				errors[addr] = err
+				// Wait for one second and no need wait after last attempt.
+				if i < 2 {
+					time.Sleep(1 * time.Second)
+				}
+			}
+		}(addr, client)
+	}
+	return errors
```
Comment (vadmeste) [inline cmd/notification.go]: wg.Wait() missing here

File: cmd/notification.go [hunk 7]
```diff
@@ -171,6 +173,174 @@ func (sys *NotificationSys) LoadCredentials() map[xnet.Host]error {
...
+					break
+				}
+				errors[addr] = err
+				// Wait for one second and no need wait after last attempt.
+				if i < 2 {
+					time.Sleep(1 * time.Second)
+				}
+			}
+		}(addr, client)
+	}
```
Comment (vadmeste) [inline cmd/notification.go:297]: wg.Wait() missing here

File: cmd/notification.go [hunk 8]
```diff
@@ -171,6 +173,174 @@ func (sys *NotificationSys) LoadCredentials() map[xnet.Host]error {
...
+					logger.LogIf(ctx, err)
+				}
+				// Wait for one second and no need wait after last attempt.
+				if i < 2 {
+					time.Sleep(1 * time.Second)
+				}
+			}
+		}(addr, client)
+	}
+	return serverInfo
```
Comment (vadmeste) [inline cmd/notification.go:340]: wg.Wait() missing here

File: cmd/notification.go [hunk 9]
```diff
@@ -171,6 +173,174 @@ func (sys *NotificationSys) LoadCredentials() map[xnet.Host]error {
...
+	var wg sync.WaitGroup
+	for addr, client := range sys.peerRPCClientMap {
+		wg.Add(1)
+		go func(addr xnet.Host, client *PeerRPCClient) {
+			defer wg.Done()
+			// Try to load credentials in three attempts.
+			for i := 0; i < 3; i++ {
+				err := client.StartProfiling(profiler)
+				if err == nil {
+					break
```
Comment (vadmeste) [inline cmd/notification.go]: The retry here has a problem here, if we encounter an error in the first try, but err == nil in the second try, the error will still be there in maps

File: cmd/admin-handlers.go [hunk 10]
```diff
@@ -359,31 +368,25 @@ func (a adminAPIHandlers) StartProfilingHandler(w http.ResponseWriter, r *http.R
...
-		go func(idx int, peer adminPeer) {
-			defer wg.Done()
-			result := StartProfilingResult{NodeName: peer.addr}
-			if err := peer.cmdRunner.StartProfiling(profiler); err != nil {
-				result.Error = err.Error()
-				return
-			}
+	var startProfilingResult []StartProfilingResult
+
+	for host, err := range globalNotificationSys.StartProfiling(profiler) {
```
Comment (vadmeste) [inline cmd/admin-handlers.go]: Currently, globalNotificationSys.StartProfiling() can be empty if there is no errors with any of the nodes.

**Review by vadmeste — DISMISSED** [review]
Some issues need to be addressed

File: cmd/admin-handlers.go (hunk 10 above)
Comment (harshavardhana) [reply cmd/admin-handlers.go]: that is a not a problem if its empty, I am not sure what you are saying.

File: cmd/notification.go (hunk 9 above)
Comment (harshavardhana) [reply cmd/notification.go]: Ah, good catch.

File: cmd/notification.go (hunk 8 above)
Comment (harshavardhana) [reply cmd/notification.go:340]: Oh yes.

File: cmd/notification.go (hunk 7 above)
Comment (harshavardhana) [reply cmd/notification.go:297]: oh yes

File: cmd/notification.go (hunk 6 above)
Comment (harshavardhana) [reply cmd/notification.go]: yes

File: cmd/admin-handlers.go (hunk 10 above)
Comment (harshavardhana) [reply cmd/admin-handlers.go]: yes this is not a problem `range nil` doesn't crash @vadmeste

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by vadmeste — COMMENTED** [review]
@harshavardhana  when downloading profiling data, only one node has non zero profiling data (mc admin profile stop myminio/; unzip -l profile.zip)

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by harshavardhana** [comment]
Ping @balamurugana @vadmeste can you review?

**Comment by minio-ops** [comment]
```
Error running mint automation
+ mkdir -p 7031-081813d/gopath/src/github.com/minio
+ git clone --quiet https://github.com/minio/minio.git 7031-081813d/gopath/src/github.com/minio/minio
++ cd 7031-081813d/gopath/src/github.com/minio/minio
++ git remote add minio https://github.com/minio/minio.git
++ cd 7031-081813d/gopath/src/github.com/minio/minio
++ git fetch --quiet minio pull/7031/head:pr7031
++ cd 7031-081813d/gopath/src/github.com/minio/minio
++ git checkout --quiet pr7031
++ minikube ip
There is a newer version of minikube available (v0.32.0).  Download it here:
https://github.com/kubernetes/minikube/releases/tag/v0.32.0

To disable this notification, run the following:
minikube config set WantUpdateNotification false
+ minikube_ip=192.168.39.88
++ mkdir -p certs
++ cd certs
++ ./cert -host 192.168.39.88
2019/01/10 11:07:33 wrote public.crt
2019/01/10 11:07:33 wrote private.key
+ GOPATH=/mnt/nvme0n1p1/mint-auto/mint-auto/7031-081813d/gopath
+ make -C 7031-081813d/gopath/src/github.com/minio/minio --quiet
# github.com/minio/minio/cmd
cmd/peer-rpc-client.go:168:1: missing return at end of function
make: *** [build] Error 2

```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by harshavardhana** [comment]
Can I get some reviews @balamurugana @vadmeste ?

**Comment by vadmeste** [comment]
@harshavardhana mc admin profile stop will download multiple profiling files, each file corresponds to one node, but all of those files except one are empty.

1. Start a distributed setup
2. `mc admin profile start myminio/`
3. `mc admin profile stop myminio/`
4. `unzip -l profile.zip`

**Comment by harshavardhana** [comment]
> @harshavardhana mc admin profile stop will download multiple profiling files, each file corresponds to one node, but all of those files except one are empty.
> 
> 1. Start a distributed setup
> 2. `mc admin profile start myminio/`
> 3. `mc admin profile stop myminio/`
> 4. `unzip -l profile.zip`

@vadmeste thanks should be fixed in the latest update to the PR

**Comment by harshavardhana** [comment]
@balamurugana can you review/approve?

**Review by vadmeste — APPROVED** [review]
LGTM & tested

**Comment by harshavardhana** [comment]
@balamurugana can you approve again?

**Review by balamurugana — APPROVED** [review]

**Pull request merged**